topdown/json.go

```diff
@@ -237,288 +241,154 @@ func pathsToObject(paths []ast.Ref) ast.Object {
 	return root
 }
 
-// toIndex tries to convert path elements (that may be strings) into indices into
-// an array.
-func toIndex(arr *ast.Array, term *ast.Term) (int, error) {
-	i := 0
+type jsonPatch struct {
+	op    string
+	path  *ast.Term
+	from  *ast.Term
+	value *ast.Term
+}
+
+func getPatch(o ast.Object) (jsonPatch, error) {
+	validOps := [6]string{"add", "remove", "replace", "move", "copy", "test"}
+	var out jsonPatch
 	var ok bool
-	switch v := term.Value.(type) {
-	case ast.Number:
-		if i, ok = v.Int(); !ok {
-			return 0, fmt.Errorf("Invalid number type for indexing")
-		}
-	case ast.String:
-		if v == "-" {
-			return arr.Len(), nil
-		}
-		num := ast.Number(v)
-		if i, ok = num.Int(); !ok {
-			return 0, fmt.Errorf("Invalid string for indexing")
+	getAttribute := func(attr string) (*ast.Term, error) {
+		if term := o.Get(ast.StringTerm(attr)); term != nil {
+			return term, nil
 		}
-		if v != "0" && strings.HasPrefix(string(v), "0") {
-			return 0, fmt.Errorf("Leading zeros are not allowed in JSON paths")
-		}
-	default:
-		return 0, fmt.Errorf("Invalid type for indexing")
-	}
-
-	return i, nil
-}
 
-// patchWorkerris a worker that modifies a direct child of a term located
-// at the given key.  It returns the new term, and optionally a result that
-// is passed back to the caller.
-type patchWorker = func(parent, key *ast.Term) (updated, result *ast.Term)
-
-func jsonPatchTraverse(
-	target *ast.Term,
-	path ast.Ref,
-	worker patchWorker,
-) (*ast.Term, *ast.Term) {
-	if len(path) < 1 {
-		return nil, nil
+		return nil, fmt.Errorf("missing '%s' attribute", attr)
 	}
 
-	key := path[0]
-	if len(path) == 1 {
-		return worker(target, key)
+	opTerm, err := getAttribute("op")
+	if err != nil {
+		return out, err
+	}
+	op, ok := opTerm.Value.(ast.String)
+	if !ok {
+		return out, fmt.Errorf("attribute 'op' must be a string")
+	}
+	out.op = string(op)
+	validOp := false
+	for i := range validOps {
+		if out.op == validOps[i] {
+			validOp = true
+			break
+		}
+	}
+	if !validOp {
+		out.op = ""
```

Comment:
Should we still return `out` when there is an error? If not, this cleanup of `out.op` can be omitted.

Reply:
You would think that we should just return `undefined` and an error, and be done with things, but the original builtin returns the unaltered original value (which unfortunately *is* a reasonable implementation decision, according to the [JSON-Patch spec](https://datatracker.ietf.org/doc/html/rfc6902/#section-5) :shrug: ).